# Description
The purpose of this notebook is to explore the ISSP dataset, identify variables of interest, and document missing values/outliers.

# Summary
## Important notes
Number of useable rows (from U.S. and has job satisfaction label): **3656**

Number of ISCO-2008 codes that need manual mapping: **36**

## Variables of interest (applicable and not too many missing values)

### Demographic
- `sex`
- `age`
- `martial_status`
- `educ_years`
- `educ_level`
- `work_status_ever`
- `household_size`
- `child_in_household`
- `religion_group`
- `urban_rural`
- `household_income_group`

### Job Value
- `job_just_for_money`
- `would_work_without_need`
- `imp_job_security`
- `imp_income`      
- `imp_advancement`
- `imp_interesting`
- `imp_independence`
- `imp_help_others`
- `imp_useful_society`
- `imp_decide_hours`    
- `pref_hours_vs_earnings`

In [1]:
import pandas as pd

DATA_DIR = "../../data/"
DATASET_DIR = DATA_DIR + "constructed_datasets/"
ISSP = DATASET_DIR + "issp.csv"

US_CODE = 840

df = pd.read_csv(ISSP)
df.head()

,study_no,archive_version,doi,case_id,cumu_id,module_year,country,country_sample,country_sample_year,time_paid_job,...,voted_last_election,religious_attendance,religion_group,social_status_scale,subjective_class,urban_rural,household_income_group,weight,data_collection_mode,ONET_SOC_CODE
0,8797.0,1.0.0 (2024-10-07),https://doi.org/10.4232/1.14391,1000001.0,2005000360000001,2005.0,36.0,36.0,362005.0,5.0,...,1.0,5.0,2.0,2.0,NaN,5.0,NaN,1.0,34.0,NaN
1,8797.0,1.0.0 (2024-10-07),https://doi.org/10.4232/1.14391,1000002.0,2005000360000002,2005.0,36.0,36.0,362005.0,1.0,...,1.0,6.0,2.0,3.0,NaN,2.0,NaN,1.0,34.0,45-1011.00
2,8797.0,1.0.0 (2024-10-07),https://doi.org/10.4232/1.14391,1000003.0,2005000360000003,2005.0,36.0,36.0,362005.0,4.0,...,1.0,2.0,2.0,6.0,NaN,1.0,2.0,1.0,34.0,NaN
3,8797.0,1.0.0 (2024-10-07),https://doi.org/10.4232/1.14391,1000004.0,2005000360000004,2005.0,36.0,36.0,362005.0,4.0,...,1.0,5.0,2.0,7.0,NaN,1.0,3.0,1.0,34.0,43-3031.00
4,8797.0,1.0.0 (2024-10-07),https://doi.org/10.4232/1.14391,1000005.0,2005000360000005,2005.0,36.0,36.0,362005.0,4.0,...,1.0,5.0,2.0,8.0,NaN,2.0,3.0,1.0,34.0,11-3071.00


In [2]:
df["job_satisfaction"].describe()

count    68929.000000
mean         2.713241
std          1.170498
min          1.000000
25%          2.000000
50%          3.000000
75%          3.000000
max          7.000000
Name: job_satisfaction, dtype: float64

In [3]:
print(f"[INFO] Potential variables:\n{"\n".join(i.strip() for i in df.columns)}")

[INFO] Potential variables:
study_no
archive_version
doi
case_id
cumu_id
module_year
country
country_sample
country_sample_year
time_paid_job
time_housework
time_family
time_friends
time_leisure
job_just_for_money
would_work_without_need
work_most_important_activity
domestic_duties
unions_needed
imp_job_security
imp_income
imp_advancement
imp_interesting
imp_independence
imp_help_others
imp_useful_society
imp_decide_hours
choice_job_kinds
choice_firm_kinds
choice_worktypes
easy_find_acceptable_job
how_hard_works
pref_hours_vs_earnings
actual_job_secure
actual_income_high
actual_advancement_high
actual_job_interesting
actual_work_independently
actual_help_others
actual_useful_society
freq_exhausted_after_work
freq_hard_physical_work
freq_work_stressful
freq_dangerous_conditions
freq_job_interferes_family
freq_family_interferes_job
relations_mgmt_employees
relations_workmates
job_satisfaction
second_job
pref_work_situation
currently_paid_work
working_hours_conditions
daily_work_organizat

# Noted variables of interest
## Demographic 
- sex
- age
- martial_status
- steady_partner
- educ_years
- educ_level
- work_status_ever
- household_size
- child_in_household
- religion_group
- social_status_scale
- subjective_class
- urban_rural
- household_income_group
- main_economic_support

## Job values
- job_just_for_money
- would_work_without_need
- work_most_important_activity
- unions_needed
- imp_job_security
- imp_income
- imp_advancement
- imp_interesting
- imp_independence
- imp_help_others
- imp_useful_society
- imp_decide_hours
- choice_job_kinds
- choice_firm_kinds
- choice_worktypes
- pref_hours_vs_earnings

# Job profile/condition variables
- actual_job_secure
- actual_income_high
- actual_advancement_high
- actual_job_interesting
- actual_work_independently
- actual_help_others
- actual_useful_society
- freq_exhausted_after_work
- freq_hard_physical_work
- freq_work_stressful
- freq_dangerous_conditions
- freq_job_interferes_family
- freq_family_interferes_job
- union_member
- second_job

## Labels and Job Code
- job_satisfaction
- isco08
- ONET_SOC_CODE

In [4]:
is_united_states = df["country"] == US_CODE
print(f"[INFO] Number of responses from U.S. market only: {df[is_united_states].shape[0]}")

[INFO] Number of responses from U.S. market only: 5676


In [5]:
# Filter DataFrame to only include rows with a job satisfaction included and in U.S.
df = df[df["job_satisfaction"].notna() & (df["country"] == US_CODE)]
print(f"[INFO] Resulting number of useable survey info (U.S. only and has job satisfaction): {df.shape[0]}")

[INFO] Resulting number of useable survey info (U.S. only and has job satisfaction): 3656


In [6]:
no_onet = df["ONET_SOC_CODE"].isna()
has_isco = df["isco08"].notna()

to_resolve = df[no_onet & has_isco]
print(f"[INFO] Number of survey responses that need O*NET code: {to_resolve.shape[0]}")

[INFO] Number of survey responses that need O*NET code: 566


In [7]:
print(f"[INFO] Number of unique ISCO-2008 codes that need mapping: {to_resolve["isco08"].nunique()}")
print(f"[INFO] Listed ISCO-2008 codes that need mapping: \n{"\n".join(str(int(i)) for i in to_resolve["isco08"].unique())}")

[INFO] Number of unique ISCO-2008 codes that need mapping: 36
[INFO] Listed ISCO-2008 codes that need mapping: 
1300
9120
5150
4100
3120
5130
2220
9210
2510
5320
5140
2160
300
3310
2640
4220
7230
2500
2210
3100
2400
2130
9300
7000
7520
8340
3420
3430
4410
5100
8180
6220
9610
5200
3210
2300


# Observation of missing codes
It looks like most of the ISCO-2008 codes that do not have a mapping are "general" codes that have not been mapped down into a specific occupation. This is likely due to the fact that it refers to a category, such as "Manager" and could not be mapped to an SOC-2010 occupation code. 

## Potential ideas to handle this
1. Since there are only 36 codes, it is not unreasonable to map this manually. 
2. Or, as our challenge advisor suggested, we could also choose a specific occupation or occupations to focus on. 
3. Finally, we could potentially map one ISCO-2008 code to multiple O\*NET-SOC 2019 code, but this will add more overhead, as we will need to decide/know what our O\*NET occupation profile dataset will look like, and have to balance that fact that a lot of our occupation data is specific to the O*NET code, so occupations under one group may disagree. 

In [8]:
with pd.option_context('display.max_rows', None):
    print("[INFO] Variables with missing values")
    print("="*50)
    na_data = df.isna().sum()
    print(na_data.loc[na_data != 0])

[INFO] Variables with missing values
time_paid_job                      1067
time_housework                     1020
time_family                         981
time_friends                        988
time_leisure                        975
job_just_for_money                   37
would_work_without_need              32
work_most_important_activity       1973
domestic_duties                    1994
unions_needed                      1885
imp_job_security                     32
imp_income                           37
imp_advancement                      37
imp_interesting                      37
imp_independence                     34
imp_help_others                      39
imp_useful_society                   36
imp_decide_hours                     48
choice_job_kinds                   1106
choice_firm_kinds                  1176
choice_worktypes                   1188
easy_find_acceptable_job           1713
how_hard_works                     2003
pref_hours_vs_earnings              182
act

# Observation of missing values
It looks like there are many variables that do not have any data for our applicable rows, such as `ever_had_paid_job_1yr`.

## Job codes
It looks like 904 rows are missing a proper ISCO-2008 code. I would suggest looking at any similar survey responses (maybe before cutting down the dataset and dropping job condition/profile features from the ISSP dataset) to try and give the survey response an appropriate ISCO-2008 code.

However, if we were to focus on a specific occupation profile, such as suggested by our challenge advisor, we might have less trouble and could focus more on other features/data engineering techniques.

## Demographic variables: missing values

### Bad variables
The following variables that were suggested above as demographic variables do not seem suitable to use, as they have too many missing values:
| Variable Name | Missing Count |
| ----------------------- | ---- |
| `steady_partner`        | 2244 |
| `social_status_scale`   | 1756 |  
| `subjective_class`      | 1952 |
| `main_economic_support` | 3656 |


## Job value variables: missing values
For many of the job value variables, it looks like there aren't too many missing. To handle these missing values, I would suggest looking into comparing between ISCO-2008 codes and seeing if there are any similar survey responses and using that to impute the missing variables.

### Good variables
See table below for missing count for job value variables:
| Variable Name | Missing Count |
| -------------------------- | -- |
| `job_just_for_money`       | 37 |
| `would_work_without_need`  | 32 |
| `imp_job_security`         | 32 |
| `imp_income`               | 37 |
| `imp_advancement`          | 37 |
| `imp_interesting`          | 37 |
| `imp_independence`         | 34 |
| `imp_help_others`          | 39 |
| `imp_useful_society`       | 36 |
| `imp_decide_hours`         | 48 |
| `pref_hours_vs_earnings`   | 182| 

### Bad variables
See table below for job value variables with too many missing values: 
| Variable Name | Missing Count | 
| ------------------------------ | ---- |
| `work_most_important_activity` | 1973 |
| `unions_needed`                | 1885 |
| `choice_job_kinds`             | 1106 |
| `choice_firm_kinds`            | 1176 |
| `choice_worktypes`             | 1188 |


In [9]:
job_val_vars = [col for col in df.columns if col.startswith("imp_")]
all_populated = df
all_populated = all_populated[all_populated[job_val_vars].notna().all(axis=1)]
all_populated.shape

(3571, 114)

In [10]:
df["job_satisfaction"].value_counts()

job_satisfaction
2.0    1291
3.0    1212
1.0     625
4.0     236
5.0     181
6.0      77
7.0      34
Name: count, dtype: int64

In [11]:
df["job_satisfaction"].describe()

count    3656.000000
mean        2.568928
std         1.205168
min         1.000000
25%         2.000000
50%         2.000000
75%         3.000000
max         7.000000
Name: job_satisfaction, dtype: float64

In [12]:
DIVIDER = "="*50
for var in job_val_vars:
    print(DIVIDER)
    print(f"[INFO] {var} description")
    print(DIVIDER)
    print(df[var].describe())
    print()

[INFO] imp_job_security description
count    3624.000000
mean        1.456954
std         0.658976
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         5.000000
Name: imp_job_security, dtype: float64

[INFO] imp_income description
count    3619.000000
mean        1.982316
std         0.777458
min         1.000000
25%         1.000000
50%         2.000000
75%         2.000000
max         5.000000
Name: imp_income, dtype: float64

[INFO] imp_advancement description
count    3619.000000
mean        1.714286
std         0.737260
min         1.000000
25%         1.000000
50%         2.000000
75%         2.000000
max         5.000000
Name: imp_advancement, dtype: float64

[INFO] imp_interesting description
count    3619.000000
mean        1.560652
std         0.652299
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         5.000000
Name: imp_interesting, dtype: float64

[INFO] imp_independence description
coun

# Observations
It looks like most of the job value variables are ranked from a scale of 1-5. 

According to the codebook, 1 is "strongly agree" and 5 is "strongly disagree", meaning that the value of the variable corresponds inversely with how important  surveyee feels the variable is.